from GetData import *
from scipy.stats import pearsonr, spearmanr

def GetAlternativeSplicingData(AS='RI', BarcodesOfInt=[]):
	DataDir = '/labs/ccurtis2/tilk/09_PROTEIN/splicingAnalysis/spliceSeq/newDownload/'
	ListOfFiles = glob.glob(os.path.join(DataDir, "PSI_download_*zip"))
	OutDF = pd.DataFrame()
	DataColumns = ['symbol', 'as_id', 'splice_type', 'exons', 'from_exon', 'to_exon']
	# Read and merge alternative splicing calls for each cancer type
	for FileName in ListOfFiles:
		CancerType = pd.read_csv(FileName, sep='\t')
		CancerType = CancerType[CancerType['splice_type'] == AS] # Select AS call types (i.e intron retention, etc)
		if len(BarcodesOfInt) == 0: # Don't subset, choose all barcodes in data frame
			OverlappingBarcodes = CancerType.columns[CancerType.columns.str.contains('TCGA')].tolist()
		else: # Choose barcodes of interest that overlap data frame
			OverlappingBarcodes = pd.Series(CancerType.columns)[pd.Series(CancerType.columns).isin(BarcodesOfInt)].tolist()
			if len(OverlappingBarcodes) == 0: # No barcodes match for this cancer type
				continue
		CancerType = CancerType[OverlappingBarcodes + DataColumns]
		if (len(OutDF) == 0):
			OutDF = CancerType
		else:
			OutDF = OutDF.merge(CancerType, left_on=DataColumns, right_on=DataColumns, how='outer')
		print('Done analyzing: ' + FileName)
	OutDF = OutDF.set_index(DataColumns)
	return(OutDF)


def GetAvgTranscriptLevelsForAS(Dataset='TCGA'):
    Exp = GetExpressionData(Dataset)
    Tissue = GetTissueType(Dataset)
    Exp = Exp.merge(Tissue, left_on='Barcode', right_on='Barcode')
    Avg = Exp.groupby(['GeneName','type'])['Value'].agg(['mean','std']).rename(columns={'mean':'AvgValue'}).reset_index()# Avg gene expression for each cancer type
    Exp = Exp.merge(Avg, left_on=['GeneName','type'], right_on=['GeneName','type'])
    Exp['DeviationFromMean'] = Exp['Value'] - Exp['AvgValue']
    Exp['DepletedTranscript'] = (Exp['DeviationFromMean'] < 0) & (Exp['DeviationFromMean'].abs() > Exp['std'])
    Exp['OverExpressedTranscript'] = (Exp['DeviationFromMean'] > 0) & (Exp['DeviationFromMean'].abs() > Exp['std'])
    Exp.to_csv('/labs/ccurtis2/tilk/scripts/protein/Data/AS_Tables/' + Dataset + '_AvgGeneExpForAllCancerTypes')
    


def GetExpLevelsForASEvents(ASType, PSI_Threshold):
    OutDir = '/labs/ccurtis2/tilk/scripts/protein/Data/AS_Tables/'
    Exp = pd.read_csv('/labs/ccurtis2/tilk/scripts/protein/Data/AS_Tables/TCGA_AvgGeneExpForAllCancerTypes') #Generated by GetAvgTranscriptLevelsForAS(Dataset='TCGA')
    Exp = Exp.merge(GetPointMutations('TCGA'), left_on=['Barcode','GeneName'], right_on=['Barcode','Hugo_Symbol'], how='left')
    Exp['Hugo_Symbol'] = Exp['Hugo_Symbol'].replace(np.nan, 0) #check that no point mutations are within the same gene
    Exp = Exp[Exp['Hugo_Symbol'] == 0]
    Exp['Barcode'] = Exp['Barcode'].str.replace('-','_').str[0:12]
    AS = GetAlternativeSplicingData(ASType).stack().reset_index().rename(columns={0:'PSI','level_6':'Barcode','symbol':'GeneName'})
    Out = AS.merge(Exp[['Barcode','GeneName','DeviationFromMean','DepletedTranscript','OverExpressedTranscript']],
        left_on=['Barcode','GeneName'], right_on=['Barcode','GeneName'])
    Out = pd.concat([Out[Out['PSI'] < PSI_Threshold].assign(Group='Not_Spliced_In'), Out[Out['PSI'] >= PSI_Threshold].assign(Group='Spliced_In')])
    drivers = pd.read_csv('/labs/ccurtis2/tilk/scripts/cancer-HRI/Data/GeneSets/drivers_Bailey2018.txt', header=None)[0]
    Out['Driver'] = Out['GeneName'].isin(drivers)
    Samples = pd.read_csv(os.getcwd() + '/Data/Raw/Mutations/TCGA/CancerTypesAndMutLoadForDGE')
    Samples['Barcode'] = Samples['Barcode'].str[0:12].str.replace('-','_')
    Out = Out.merge(Samples, left_on='Barcode', right_on='Barcode', how='left')
    Out['NMD'] = (Out['Group'] == 'Spliced_In') & (Out['DepletedTranscript'] == True)
    Out.groupby(['Barcode','MutLoad','NMD']).size().unstack().reset_index().replace(np.nan,0).to_csv(
        OutDir + 'TCGA_'+ ASType +'_Counts_ThresholdByPSI_' + str(PSI_Threshold))
    Out.groupby(['Barcode','MutLoad','Driver','NMD']).size().unstack().reset_index().replace(np.nan,0).to_csv(
        OutDir + 'TCGA_'+ ASType +'_Counts_ThresholdByPSI_' + str(PSI_Threshold) + '_ByDrivers_')
    print('Done!')
    return(Out)




def DoTTestForAS(row):
    '''This is a two-sided test for the null hypothesis that 2 independent samples have identical average (expected) values.'''
    Samples = pd.read_csv(os.getcwd() + '/Data/Raw/Mutations/TCGA/CancerTypesAndMutLoadForDGE')
    Samples['Barcode'] = Samples['Barcode'].str[0:12].str.replace('-','_')
    Low = row[list(set(row.index) & set(Samples[Samples['MutLoad'] <= 10]['Barcode'].tolist()))].dropna()
    High = row[list(set(row.index) & set(Samples[Samples['MutLoad'] >= 1000]['Barcode'].tolist()))].dropna()
    return( pd.Series({'pVal': stats.ttest_ind(Low, High)[1], 
						'Low_mean': Low.mean(), 
						'High_mean': High.mean(), 
						'LowToHighDelta': Low.mean() - High.mean()}))


def GetCorCoef(df):
    df = df.dropna()
    try:
        Cor = pearsonr(df['Bin'], df['Fraction'])
        return(pd.Series({'Cor': Cor[0], 'P-val':Cor[1]}))
    except:
        return(pd.Series({'Cor':np.nan,'P-val':np.nan}))
    


    


thresh= 0.5


GetExpLevelsForASEvents(ASType='RI', PSI_Threshold=thresh)
GetExpLevelsForASEvents(ASType='ES', PSI_Threshold=thresh)
GetExpLevelsForASEvents(ASType='AD', PSI_Threshold=thresh)
GetExpLevelsForASEvents(ASType='AP', PSI_Threshold=thresh)
GetExpLevelsForASEvents(ASType='AT', PSI_Threshold=thresh)
GetExpLevelsForASEvents(ASType='AA', PSI_Threshold=thresh)
GetExpLevelsForASEvents(ASType='ME', PSI_Threshold=thresh)
